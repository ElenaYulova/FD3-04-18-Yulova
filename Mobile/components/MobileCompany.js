import React from 'react';
import PropTypes from 'prop-types';

import MobileClient from './MobileClient';

import './MobileCompany.css';

class MobileCompany extends React.PureComponent {

  static propTypes = {
    name: PropTypes.string.isRequired,
    clients: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.number.isRequired,
        fam: PropTypes.string.isRequired,
        im: PropTypes.string.isRequired,
        otch: PropTypes.string.isRequired,
        balance: PropTypes.number.isRequired,
        shown: PropTypes.bool,
      })
    ),
  };

  state = {
    name: this.props.name,
    clients: this.props.clients.map((v) => {
      v.edited = false;
      return v;
    }),
    notFilteredClients: this.props.clients.map((v) => {
      v.edited = false;
      return v;
    }),
  };

  setName = (EO) => {
    this.setState({name: EO.target.value});
  };

 
  
  setBalance = (clientId,newBalance) => {
    let changed=false;
    let newClients=[...this.state.clients]; // копия самого массива клиентов
    let newClient;
    let newFilteredClients = [...this.state.notFilteredClients];

    newClients.forEach( (c,i) => {
      if ( c.id==clientId ) {
        newClient={...c}; // копия хэша изменившегося клиента
        newClient.balance=newBalance;
        newClients[i]=newClient;
        changed=true;
      }
    } );
    if ( changed ) {
      newFilteredClients.forEach((c, i) => {
        if (c.id == clientId) {
          newFilteredClients[i] = newClient;
        }
      });
      this.setState({
        clients: newClients,
        notFilteredClients: newFilteredClients,
      });
    } 
    };
  setClientFIO  = (clientId,newFIO) => {
    let changed=false;
    let newClients=[...this.state.clients]; // копия самого массива клиентов
    let newClient;
    let newFilteredClients = [...this.state.notFilteredClients];
    newClients.forEach( (c,i) => {
      if (c.id == clientId) {
        newClient = { ...c };
        newClient.edited = false;

        if (c.fam != newFIO.fam || c.im != newFIO.im || c.otch != newFIO.otch) {
            newClient.fam = newFIO.fam;
            newClient.im = newFIO.im;
            newClient.otch = newFIO.otch;
        }
        newClients[i] = newClient;
        changed = true;
      }
    } );
    if (changed) {
      newFilteredClients.forEach((c, i) => {
        if (c.id == clientId) {
          newFilteredClients[i] = newClient;
        }
      });

      this.setState({
        clients: newClients,
        notFilteredClients: newFilteredClients,
      });
    }
  };
  addClient = () => {
    let currentClients = this.state.clients;
    let newClientId = currentClients[currentClients.length - 1].id + 1;
    let newClients = [...this.state.clients];
    let newFilteredClients = [...this.state.notFilteredClients];
    let newClient = {
      id: newClientId,
      fam: "",
      im: "",
      otch: "",
      balance: this.state.filter == "blocked" ? -1 : 0,
      edited: true,
    }
    newClients.push(newClient);
    newFilteredClients.push(newClient);
    this.setState({ 
      clients: newClients,
      notFilteredClients: newFilteredClients });
    // console.log(this.state.notFilteredClients);
  } 

  deleteClient = (clientId) => {
    let newClients=[...this.state.cliens];
    let newFilteredClients = [...this.state.notFilteredClients];
    newClients.forEach((c, i) => {
      if (c.id == clientId) {
        newClients.splice(i, 1);
        return;
      }
    });
    newFilteredClients.forEach((c, i) => {
      if (c.id == clientId) {
        newFilteredClients.splice(i, 1);
        return;
      }
    });

    this.setState({
      clients: newClients,
      notFilteredClients: newFilteredClients,
    });
  };
  
  clientFilter = (EO) => {
    var newFilter = EO.target.value;
    if (newFilter !== this.state.filter) {
      let clients = this.state.clients;
      let filteredClients = [];

      if (newFilter == "all") {
        filteredClients = [...clients];
      } else if (newFilter == "active") {
        filteredClients = clients.filter((c) => c.balance >= 0);
      } else if (newFilter == "blocked") {
        filteredClients = clients.filter((c) => c.balance < 0);
      }

      this.setState({
        filter: newFilter,
        notFilteredClients: filteredClients,
      });
      console.log(this.state.notFilteredClients);
    }
  }
  
  
  render() {

    console.log("MobileCompany render");

    var clients=this.state.notFilteredClients.map(client => {
        
        return <MobileClient 
        key={client.id} 
        client={client}
        onFIOChange={this.setClientFIO}
        onBalanceChange={this.setBalance}
        onDelete={this.deleteClient}
         />;
      }
    );

    return (
      <div className='MobileCompany'>
         <div className="MobileCompanyTitle">
          <div>
            <input type="button" value="Velcom" onClick={this.setName} />
            <input type="button" value="MTC" onClick={this.setName} />
          </div>
          <div className="">
          <label htmlFor="filter">Show</label>
            <select id="filter" value={this.state.filter} onChange={this.clientFilter}>
              <option value="all">All</option>
              <option value="active">Active</option>
              <option value="blocked">Blocked</option>
            </select>
          </div>
        </div>
        <div className='MobileCompanyName'>Компания &laquo;{this.state.name}&raquo;</div>
        <div className='MobileCompanyClients'>
          {clients}
        </div>
        <button className="MobileNewClientButton" onClick={this.addClient}>
          Add new client
        </button>
      </div>
    )
    ;

  }

}

export default MobileCompany;
