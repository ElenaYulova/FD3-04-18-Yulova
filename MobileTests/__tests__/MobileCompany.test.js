

"use strict";

import React from 'react';
import renderer from 'react-test-renderer';

import {shallow, configure} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import MobileCompany from '../components/MobileCompany';

configure({ adapter: new Adapter() });

describe('>>>MobileCompany-- Shallow Render REACT COMPONENTS', () => {
//shapshots
    let companyName='Velcom';
    let clientsArr=[ 
    {id:101, fam:"Иванов", im:"Иван", otch:"Иванович", balance:200}, 
    {id:105, fam:"Сидоров", im:"Сидор", otch:"Сидорович", balance:250}, 
    {id:110, fam:"Петров", im:"Пётр", otch:"Петрович", balance:-1},
    {id:120, fam:"Григорьев", im:"Григорий", otch:"Григорьевич", balance:220},
    ];

    const companyComponent = <MobileCompany name={companyName} clients={clientsArr} />;
    const component = renderer.create(companyComponent);

    it('render company', () => {
        let componentTree = component.toJSON();
        expect(componentTree).toMatchSnapshot();
      });
    
      let wrapper;
      beforeEach(() => {
        wrapper = shallow(companyComponent);
      });

    //setClientBalance
    it('set balance', () => {
        wrapper.instance().setBalance(101, -1);
        expect(wrapper.state('clients')[0].balance).toEqual(-1);
      });

    //edit FIO
    it('set FIO', () => {
        wrapper.instance().setClientFIO(101, {
          fam: 'fam',
          im: 'im',
          otch: 'otch',
        });
        expect(wrapper.state('clients')[0].fam).toEqual('fam');
        expect(wrapper.state('clients')[0].im).toEqual('im');
        expect(wrapper.state('clients')[0].otch).toEqual('otch');
      });

    //add new client
    it('add new client', () => {
        wrapper.instance().addClient();
        const clients = wrapper.state('clients');
        expect(clients.length).toEqual(5);
        expect(clients[clients.length-1].edited).toEqual(true);
      });

    //delete client
    it('delete client', () => {
        wrapper.instance().deleteClient("101");
        expect(wrapper.state('clients').length).toEqual(3);
       });

    //filter clients
      it('filter clients', () => {
        const event = {
          target: {value: 'active'}
        }
    
        wrapper.instance().clientFilter(event);
        expect(wrapper.state('notFilteredClients').length).toEqual(3);
        
        event.target.value = 'blocked';
        wrapper.instance().clientFilter(event);
        expect(wrapper.state('notFilteredClients').length).toEqual(1);
      });
    
    

});