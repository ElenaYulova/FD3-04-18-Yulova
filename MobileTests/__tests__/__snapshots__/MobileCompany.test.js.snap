// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>>MobileCompany-- Shallow Render REACT COMPONENTS render company 1`] = `
<div
  className="MobileCompany"
>
  <div
    className="MobileCompanyTitle"
  >
    <div>
      <input
        onClick={[Function]}
        type="button"
        value="Velcom"
      />
      <input
        onClick={[Function]}
        type="button"
        value="MTC"
      />
    </div>
    <div
      className=""
    >
      <label
        htmlFor="filter"
      >
        Show
      </label>
      <select
        id="filter"
        onChange={[Function]}
        value={undefined}
      >
        <option
          value="all"
        >
          All
        </option>
        <option
          value="active"
        >
          Active
        </option>
        <option
          value="blocked"
        >
          Blocked
        </option>
      </select>
    </div>
  </div>
  <div
    className="MobileCompanyName"
  >
    Компания «
    Velcom
    »
  </div>
  <div
    className="MobileCompanyClients"
  >
    <div
      className="MobileClient"
    >
      <div
        className="MobileClient"
      >
        <span
          className="MobileClientId"
        >
          101
        </span>
        <div
          className="MobileClientFio"
        >
          <div
            className="MobileClientFioContainer"
          >
            <span>
              Иванов Иван Иванович
            </span>
            <button
              onClick={[Function]}
            >
              Edit
            </button>
          </div>
        </div>
        <div>
          <span
            className="MobileClientBalance"
          >
            200
          </span>
          <button
            className="decrease"
            onClick={[Function]}
          >
            Decrease
          </button>
          <button
            className="increase"
            onClick={[Function]}
          >
            Increase
          </button>
        </div>
        <div
          className="MobileClientActive"
        >
          <span
            className="MobileClientActiveActive"
          >
            Active
          </span>
        </div>
        <div
          className="MobileClientDelete"
        >
          <button
            className="delete"
            onClick={[Function]}
          >
            Delete
          </button>
        </div>
      </div>
    </div>
    <div
      className="MobileClient"
    >
      <div
        className="MobileClient"
      >
        <span
          className="MobileClientId"
        >
          105
        </span>
        <div
          className="MobileClientFio"
        >
          <div
            className="MobileClientFioContainer"
          >
            <span>
              Сидоров Сидор Сидорович
            </span>
            <button
              onClick={[Function]}
            >
              Edit
            </button>
          </div>
        </div>
        <div>
          <span
            className="MobileClientBalance"
          >
            250
          </span>
          <button
            className="decrease"
            onClick={[Function]}
          >
            Decrease
          </button>
          <button
            className="increase"
            onClick={[Function]}
          >
            Increase
          </button>
        </div>
        <div
          className="MobileClientActive"
        >
          <span
            className="MobileClientActiveActive"
          >
            Active
          </span>
        </div>
        <div
          className="MobileClientDelete"
        >
          <button
            className="delete"
            onClick={[Function]}
          >
            Delete
          </button>
        </div>
      </div>
    </div>
    <div
      className="MobileClient"
    >
      <div
        className="MobileClient"
      >
        <span
          className="MobileClientId"
        >
          110
        </span>
        <div
          className="MobileClientFio"
        >
          <div
            className="MobileClientFioContainer"
          >
            <span>
              Петров Пётр Петрович
            </span>
            <button
              onClick={[Function]}
            >
              Edit
            </button>
          </div>
        </div>
        <div>
          <span
            className="MobileClientBalance"
          >
            -1
          </span>
          <button
            className="decrease"
            onClick={[Function]}
          >
            Decrease
          </button>
          <button
            className="increase"
            onClick={[Function]}
          >
            Increase
          </button>
        </div>
        <div
          className="MobileClientActive"
        >
          <span
            className="MobileClientActiveBlocked"
          >
            Blocked
          </span>
        </div>
        <div
          className="MobileClientDelete"
        >
          <button
            className="delete"
            onClick={[Function]}
          >
            Delete
          </button>
        </div>
      </div>
    </div>
    <div
      className="MobileClient"
    >
      <div
        className="MobileClient"
      >
        <span
          className="MobileClientId"
        >
          120
        </span>
        <div
          className="MobileClientFio"
        >
          <div
            className="MobileClientFioContainer"
          >
            <span>
              Григорьев Григорий Григорьевич
            </span>
            <button
              onClick={[Function]}
            >
              Edit
            </button>
          </div>
        </div>
        <div>
          <span
            className="MobileClientBalance"
          >
            220
          </span>
          <button
            className="decrease"
            onClick={[Function]}
          >
            Decrease
          </button>
          <button
            className="increase"
            onClick={[Function]}
          >
            Increase
          </button>
        </div>
        <div
          className="MobileClientActive"
        >
          <span
            className="MobileClientActiveActive"
          >
            Active
          </span>
        </div>
        <div
          className="MobileClientDelete"
        >
          <button
            className="delete"
            onClick={[Function]}
          >
            Delete
          </button>
        </div>
      </div>
    </div>
  </div>
  <button
    className="MobileNewClientButton"
    onClick={[Function]}
  >
    Add new client
  </button>
</div>
`;
